pkg/goformat/runner.go

```diff
@@ -0,0 +1,166 @@
+
+func NewRunner(log logutils.Log,
+	metaFormatter *goformatters.MetaFormatter, matcher *processors.GeneratedFileMatcher,
+	opts RunnerOptions) *Runner {
+	return &Runner{
+		log:           log,
+		matcher:       matcher,
+		metaFormatter: metaFormatter,
+		opts:          opts,
+	}
+}
+
+func (c *Runner) Run(paths []string) error {
+	for _, path := range paths {
+		err := c.walk(path)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Runner) walk(root string) error {
+	return filepath.Walk(root, func(path string, f fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if f.IsDir() && skipDir(f.Name()) {
+			return fs.SkipDir
+		}
+
+		// Ignore non-Go files.
+		if !isGoFile(f) {
+			return nil
+		}
+
+		match, err := c.opts.MatchPatterns(path)
+		if err != nil || match {
+			return err
+		}
+
+		input, err := os.ReadFile(path)
+		if err != nil {
+			return err
+		}
+
+		match, err = c.matcher.IsGeneratedFile(path, input)
+		if err != nil || match {
+			return err
+		}
+
+		output := c.metaFormatter.Format(path, input)
+
+		if bytes.Equal(input, output) {
+			return nil
+		}
+
+		c.log.Infof("format: %s", path)
+
+		// On Windows, we need to re-set the permissions from the file. See golang/go#38225.
+		var perms os.FileMode
+		if fi, err := os.Stat(path); err == nil {
+			perms = fi.Mode() & os.ModePerm
+		}
+
+		return os.WriteFile(path, output, perms)
+	})
+}
+
+type RunnerOptions struct {
+	basePath  string
+	patterns  []*regexp.Regexp
+	generated string
+}
+
+func NewRunnerOptions(cfg *config.Config) (RunnerOptions, error) {
+	basePath, err := fsutils.GetBasePath(context.Background(), cfg.Run.RelativePathMode, cfg.GetConfigDir())
+	if err != nil {
+		return RunnerOptions{}, fmt.Errorf("get base path: %w", err)
+	}
+
+	opts := RunnerOptions{
+		basePath:  basePath,
+		generated: cfg.Formatters.Exclusions.Generated,
+	}
+
+	for _, pattern := range cfg.Formatters.Exclusions.Paths {
+		exp, err := regexp.Compile(fsutils.NormalizePathInRegex(pattern))
+		if err != nil {
+			return RunnerOptions{}, fmt.Errorf("compile path pattern %q: %w", pattern, err)
+		}
+
+		opts.patterns = append(opts.patterns, exp)
+	}
+
+	return opts, nil
+}
+
+func (o RunnerOptions) MatchPatterns(path string) (bool, error) {
+	if len(o.patterns) == 0 {
+		return false, nil
+	}
+
+	rel, err := filepath.Rel(o.basePath, path)
+	if err != nil {
+		return false, err
+	}
+
+	for _, pattern := range o.patterns {
+		if pattern.MatchString(rel) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func skipDir(name string) bool {
+	switch name {
+	case "vendor", "testdata", "node_modules":
+		return true
+
+	case "third_party", "builtin": // For compatibility with `exclude-dirs-use-default`.
+		return true
+
+	default:
+		if strings.HasPrefix(name, ".") {
+			return true
+		}
+
+		return false
+	}
+}
+
+func isGoFile(f fs.FileInfo) bool {
+	return !f.IsDir() && !strings.HasPrefix(f.Name(), ".") && strings.HasSuffix(f.Name(), ".go")
```

Comment:
Can we use `filepath.Ext(f.Name()) == ".go"` instead of `strings.HasSuffix(f.Name(), ".go")`?

Reply:
https://github.com/golang/go/blob/e81f7155154c0f5d40363e84a8f24a5b559b5eed/src/cmd/fix/main.go#L264